### PR TITLE
store/tikv: Make GC's tests more strict

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve? 

This PR tries to make GCWorker's tests better.

### What is changed and how it works?

This PR does following changes:
- [x] Implement GC to Mock TiKV
- [x] Tests for GC of Mock TiKV
- [x] Check whether data is really GCed in existing tests
- [x] Add test for `leaderTick` function
- [x] Cover some error cases

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Increased code complexity

### CODE REVIEW

**Comment by MyonKeminta** [comment]
@disksing @AndreMouche PTAL

File: store/mockstore/mocktikv/pd.go [hunk 1]
```diff
@@ -108,7 +110,10 @@ func (c *pdClient) GetAllStores(ctx context.Context, opts ...pd.GetStoreOption)
 }
 
 func (c *pdClient) UpdateGCSafePoint(ctx context.Context, safePoint uint64) (uint64, error) {
-	return 0, nil
+	if safePoint > c.gcSafePoint {
```
Comment (disksing) [inline store/mockstore/mocktikv/pd.go]: Do we need lock here?

File: store/tikv/gcworker/gc_worker_test.go [hunk 2]
```diff
@@ -98,6 +100,82 @@ func (s *testGCWorkerSuite) timeEqual(c *C, t1, t2 time.Time, epsilon time.Durat
...
+	// safePoint is successfully uploaded.
+	safePoint, err := s.pdClient.UpdateGCSafePoint(context.Background(), 0)
+	c.Assert(err, IsNil)
+	return safePoint
+}
+
+// garbage represents a key that contains multiple versions, one of which should be collected.
+// This is not used to check the correctness of GC algorithm, but only for checking whether GC has been executed on the
+// specified key. Create this using `s.createGarbage`.
+type garbage struct {
```
Comment (disksing) [inline store/tikv/gcworker/gc_worker_test.go]: How about `GCProbe` or `GCDetector`?

File: store/tikv/gcworker/gc_worker_test.go [hunk 3]
```diff
@@ -98,6 +100,82 @@ func (s *testGCWorkerSuite) timeEqual(c *C, t1, t2 time.Time, epsilon time.Durat
...
+	key string
+	// The ts that can see the version that should be deleted.
+	firstTs uint64
```
Comment (disksing) [inline store/tikv/gcworker/gc_worker_test.go]: How about `v1Ts`?

**Comment by MyonKeminta** [comment]
@disksing @AndreMouche  PTAL thanks!

**Comment by MyonKeminta** [comment]
@disksing  Sorry I was resolving conflicts. Please approve again and reply a "LGTM"

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 4]
```diff
@@ -1047,6 +1047,72 @@ func (mvcc *MVCCLevelDB) BatchResolveLock(startKey, endKey []byte, txnInfos map[
...
+					return errors.Trace(err)
+				}
+				break
+			}
+
+			if dec.value.commitTS > safePoint {
+				continue
+			}
+
+			if dec.value.valueType == typePut || dec.value.valueType == typeDelete {
```
Comment (AndreMouche) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: I think we should delete all version once `keepNext` is false?

Comment (MyonKeminta) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: It's exactly what L1102-1103 does, isn't it?

**Pull request merged**